### PR TITLE
feat: Add tracing to outbound oauth introspection requests

### PR DESCRIPTION
**Pull request by snikch**
## Proposed changes

Support for tracing was added in #376, however tracing isn't added for outbound requests, i.e. `oauth2_introspection` authentication.

This PR adds in tracing to the outbound request for OAuth2 Introspection, resulting in a single trace being propagated all the way through Oathkeeper and into Hydra (or whatever downstream system is being used).

### Before - two traces, no propagation

image
image

### After - single trace, propagated!

image

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by snikch** [comment]
Note: failures appear to incorrect. Windows test appears to be flaky and the circleci failure looks incorrect:

```
ERRO Running error: context loading failed: no go files to analyze 
ERRO Timeout exceeded: try increasing it by passing --timeout option 
```

File: pipeline/authn/authenticator_oauth2_introspection.go [hunk 1]
```diff
@@ -155,7 +181,14 @@ func (a *AuthenticatorOAuth2Introspection) Authenticate(r *http.Request, session
 		}
 		// set/override the content-type header
 		introspectReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+		// add tracing
+		closeSpan := a.traceRequest(r.Context(), introspectReq)
+
```
Comment (aeneasr) [inline pipeline/authn/authenticator_oauth2_introspection.go:187]: Should probably be:

```
defer a.traceRequest(r.Context(), introspectReq)
```

**Review by aeneasr — CHANGES_REQUESTED** [review]
Awesome, thank you for your contribution! This looks pretty good and I have some ideas how to improve it further :)

File: pipeline/authn/authenticator_oauth2_introspection.go (hunk 1 above)
Comment (snikch) [reply pipeline/authn/authenticator_oauth2_introspection.go:187]: If we use `defer` here then we'll be including all of the later code inside the trace span, which may lead to an inflated duration of the span. The later call `tokenToCache` - is that just an in memory cache. If so then we could use `defer`, but making the call to `closeSpan` explicit means theres zero chance of later adding code that inflates the span duration.

Comment (aeneasr) [reply pipeline/authn/authenticator_oauth2_introspection.go:187]: My bad, that makes total sense!

**Review by aeneasr — APPROVED** [review]
Awesome 🎉 

Thank you for your contribution!

**Comment by snikch** [comment]
The tests are continuing to fail. Is that anything that I've done? I thought it was a flaky test but it seems to be consistent.

**Comment by aeneasr** [comment]
[Formatting yes](https://app.circleci.com/pipelines/github/ory/oathkeeper/882/workflows/2545a85f-612d-4c6f-b2a4-4cf0664f0a5b/jobs/7977) (run `make format`), the other one seems to be a flake.

**Comment by snikch** [comment]
Ok, just the windows test failing now.

**Comment by aeneasr** [comment]
Awesome, thank you! The windows tests have been flaky for some time!

**Pull request merged**